File: sources/main/java/net/zxmushroom63/plugins/PluginLoader.js
window.pluginLoader = function pluginLoader(pluginsArr) {
  function checkPluginsLoaded(totalLoaded, identifier) {
    console.log(
      "Checking if Plugins are finished :: " +
        totalLoaded +
        "/" +
        pluginsArr.length
    );
    if (totalLoaded >= pluginsArr.length) {
      clearInterval(identifier);
      window.pluginGracePeriod = false;
      console.log(
        "Checking if Plugins are finished :: All plugins loaded! Grace period off."
      );
    }
  }
  window.pluginGracePeriod = true;
  var totalLoaded = 0;
  var loaderCheckInterval = null;
  pluginsArr.forEach((c) => {
    let currentPlugin = c;
    console.log("EaglerPL: Starting " + currentPlugin);
    fetch(currentPlugin)
      .then((x) => {
        x.blob().then((y) => {
          var reader = new FileReader();
          reader.onloadend = () => {
            console.log(
              "EaglerPL: Loading " + currentPlugin + " via method A."
            );
            var script = document.createElement("script");
            script.src = reader.result;
            script.setAttribute("data-plugin", currentPlugin);
            script.setAttribute("data-isplugin", true);
            script.onerror = () => {
              console.log(
                "EaglerPL: Failed to load " + currentPlugin + " via method A!"
              );
              script.remove();
              totalLoaded++;
            };
            script.onload = () => {
              console.log(
                "EaglerPL: Successfully loaded " +
                  currentPlugin +
                  " via method A."
              );
              totalLoaded++;
            };
            document.body.appendChild(script);
          };
          reader.readAsDataURL(y);
        });
      })
      .catch((err) => {
        try {
          console.log("EaglerPL: Loading " + currentPlugin + " via method B.");
          var script = document.createElement("script");
          script.src = currentPlugin;
          script.setAttribute("data-plugin", currentPlugin);
          script.setAttribute("data-isplugin", true);
          script.onerror = () => {
            console.log(
              "EaglerPL: Failed to load " + currentPlugin + " via method B!"
            );
            script.remove();
            totalLoaded++;
          };
          script.onload = () => {
            console.log(
              "EaglerPL: Successfully loaded " +
                currentPlugin +
                " via method B."
            );
            totalLoaded++;
          };
          document.body.appendChild(script);
        } catch (error) {
          console.log(
            "EaglePL: Oh no! The plugin " + currentPlugin + " failed to load!"
          );
          totalLoaded++;
        }
      });
  });
  loaderCheckInterval = setInterval(() => {
    checkPluginsLoaded(totalLoaded, loaderCheckInterval);
  }, 500);
  console.log("EaglerPL: Starting to load "+pluginsArr.length+" plugins...");
};

File: sources/main/java/net/zxmushroom63/plugins/PluginLoader.java
package net.zxmushroom63.plugins;

import org.teavm.jso.JSBody;

public class PluginLoader {
    public static String[] plugins = {};

    @JSBody(params = { "plugins" }, script = "window.pluginLoader(plugins);")
    public static native void loadPlugins(String[] plugins);

    @JSBody(params = {}, script = "window.pluginLoader = function pluginLoader(pluginsArr) {\r\n" + //
            "  function checkPluginsLoaded(totalLoaded, identifier) {\r\n" + //
            "    console.log(\"loadercheck :: \" + totalLoaded);\r\n" + //
            "    if (totalLoaded >= pluginsArr.length) {\r\n" + //
            "      clearInterval(identifier);\r\n" + //
            "      window.pluginGracePeriod = false;\r\n" + //
            "      console.log(\"loadercheck :: done!\");\r\n" + //
            "    }\r\n" + //
            "  }\r\n" + //
            "  window.pluginGracePeriod = true;\r\n" + //
            "  var totalLoaded = 0;\r\n" + //
            "  var loaderCheckInterval = null;\r\n" + //
            "  var i;\r\n" + //
            "  pluginsArr.forEach((c) => {\r\n" + //
            "    let currentPlugin = c;\r\n" + //
            "    console.log(\"ILOADER: DOING \" + currentPlugin);\r\n" + //
            "    fetch(currentPlugin)\r\n" + //
            "      .then((x) => {\r\n" + //
            "        x.blob().then((y) => {\r\n" + //
            "          var reader = new FileReader();\r\n" + //
            "          reader.onloadend = () => {\r\n" + //
            "            console.log(\"ILOADER: METHOD1LOADSTART \" + currentPlugin);\r\n" + //
            "            var script = document.createElement(\"script\");\r\n" + //
            "            script.src = reader.result;\r\n" + //
            "            script.setAttribute(\"data-plugin\", currentPlugin);\r\n" + //
            "            script.setAttribute(\"data-isplugin\", true);\r\n" + //
            "            script.onerror = () => {\r\n" + //
            "              console.log(\"ILOADER: METHOD1ERR \" + currentPlugin);\r\n" + //
            "              script.remove();\r\n" + //
            "              totalLoaded++;\r\n" + //
            "            };\r\n" + //
            "            script.onload = () => {\r\n" + //
            "              console.log(\"ILOADER: METHOD1LOAD \" + currentPlugin);\r\n" + //
            "              totalLoaded++;\r\n" + //
            "            };\r\n" + //
            "            document.body.appendChild(script);\r\n" + //
            "          };\r\n" + //
            "          reader.readAsDataURL(y);\r\n" + //
            "        });\r\n" + //
            "      })\r\n" + //
            "      .catch((err) => {\r\n" + //
            "        try {\r\n" + //
            "          console.log(\"ILOADER: METHOD2LOADSTART \" + currentPlugin);\r\n" + //
            "          var script = document.createElement(\"script\");\r\n" + //
            "          script.src = currentPlugin;\r\n" + //
            "          script.setAttribute(\"data-plugin\", currentPlugin);\r\n" + //
            "          script.setAttribute(\"data-isplugin\", true);\r\n" + //
            "          script.onerror = () => {\r\n" + //
            "            console.log(\"ILOADER: METHOD2ERR \" + currentPlugin);\r\n" + //
            "            script.remove();\r\n" + //
            "            totalLoaded++;\r\n" + //
            "          };\r\n" + //
            "          script.onload = () => {\r\n" + //
            "            console.log(\"ILOADER: METHOD2LOAD \" + currentPlugin);\r\n" + //
            "            totalLoaded++;\r\n" + //
            "          };\r\n" + //
            "          document.body.appendChild(script);\r\n" + //
            "        } catch (error) {\r\n" + //
            "          console.log(\"ILOADER: SAFEGUARD WAS HIT \" + currentPlugin);\r\n" + //
            "          totalLoaded++;\r\n" + //
            "        }\r\n" + //
            "      });\r\n" + //
            "  });\r\n" + //
            "  console.log(\"Done loading plugins!\");\r\n" + //
            "  loaderCheckInterval = setInterval(() => {\r\n" + //
            "    checkPluginsLoaded(totalLoaded, loaderCheckInterval);\r\n" + //
            "  }, 500);\r\n" + //
            "  console.log(\"Began loading plugins...\");\r\n" + //
            "};")
    public static native void loadLoader();

    @JSBody(params = { "plugins" }, script = "localStorage.setItem('pl::plugins', JSON.stringify(plugins))")
    private static native void savePlugins(String[] plugins);

    @JSBody(params = {}, script = "try { return JSON.parse(localStorage.getItem('pl::plugins')||'[]') } catch(err) {return []}")
    private static native String[] retrievePlugins();

    public static void savePluginsToLocalStorage() {
        savePlugins(plugins);
    };

    public static void loadPluginsFromLocalStorage() {
        plugins = retrievePlugins();
    };
}